Include/pyport.h

```diff
@@ -160,6 +160,27 @@ typedef int Py_ssize_clean_t;
 #   define PY_FORMAT_SIZE_T "z"
 #endif
 
+
+/* _Py_HAVE_TYPEOF and _Py_TYPEOF can opportunistically support an equivalent
+ * of GCC's typeof extension where possible. It is not possible to get
+ * equivalent behavior on all platforms, so all uses of _Py_TYPEOF should be
+ * guarded by _Py_HAVE_TYPEOF.
+ */
+#if defined(__GNUC__) || defined(__clang__) || defined(__cplusplus)
+#   define _Py_HAVE_TYPEOF 1
+#   if defined(__cplusplus)
+#       define _Py_TYPEOF(x) decltype(x)
```

Comment:
I'm not sure that it's worth it to add _Py_TYPEOF() just to support C++. C extensions are usually built in C. All stdlib C extensions are built in C mode.

Since it's a private macro, I suggest to only support C and documented _Py_HAVE_TYPEOF as `Define if __typeof__() builtin function is available`.

Reply:
I think the purpose here was not so much to support C++ but to give us some compiler-independent behavior. In C++ mode, there's an alternative to `typeof` that works in all the scenarios we use `__typeof__` for, so this extends the scenarios where we can do `__typeof__`-like operations.

I also think that *maybe* we want `_Py_TYPEOF()` to be defined as `__extension__ ( __typeof__(x) ) `, in which case it's better to have our own macro defined for it (and at that point the `__cplusplus` definition is only a cost in that we may not be testing it properly if we never build anything in C++ mode).

The other benefit to this is that if other compilers add support for `typeof`-like functionality, we'll be able to seamlessly add them.